src/adios4dolfinx/comm_helpers.py

```diff
@@ -245,15 +233,20 @@ def send_dofs_and_recv_values(
     dofs_offsets = np.zeros(len(out_size) + 1, dtype=np.intc)
     dofs_offsets[1:] = np.cumsum(out_size)
     out_dofs = np.zeros(dofs_offsets[-1], dtype=np.int64)
-    dof_count = np.zeros_like(out_size, dtype=np.int32)
-    proc_to_local = np.zeros_like(
-        input_dofmap, dtype=np.int32
-    )  # Map output to local dof
-    for i, (dof, owner) in enumerate(zip(input_dofmap, dofmap_owners)):
-        proc_pos = find_first(owner, dest_ranks)
-        out_dofs[dofs_offsets[proc_pos] + dof_count[proc_pos]] = dof
-        proc_to_local[dofs_offsets[proc_pos] + dof_count[proc_pos]] = i
-        dof_count[proc_pos] += 1
+    proc_to_local = np.zeros_like(input_dofmap, dtype=np.int32)
+
+    # Map output to local dof
+    proc_row, proc_col = np.nonzero(process_pos_indicator)
+    assert np.allclose(proc_row, np.arange(len(process_pos_indicator), dtype=np.int32))
+
+    cum_pos = np.cumsum(process_pos_indicator, axis=0)
+    insert_position = cum_pos[np.arange(len(proc_col), dtype=np.int32), proc_col] - 1
+    out_dofs = np.zeros(dofs_offsets[-1], dtype=np.int64)
+    insertion_array = dofs_offsets[proc_col] + insert_position
+    out_dofs[insertion_array] = input_dofmap
+    proc_to_local = np.zeros_like(input_dofmap, dtype=np.int32)
+    proc_to_local[insertion_array] = np.arange(len(input_dofmap), dtype=np.int32)
+    del insertion_array, cum_pos, insert_position
```

Comment:
This snippet seem to be repeated a couple of time. Would be nice to create a helper function to do this so that we avoid too much repetition.

Reply:
It's a bit tricky to generalize it,as each of them use slightly different input/output. I will not try to unify them for now.